### ISSUE TITLE
Investigate a crash in wazuh-db

### ISSUE THREAD

**Issue by vikman90** [wazuh/wazuh#16749]
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| 4.3.10 | wazuh-db | Manager | Packages | CentOS 7 |

We've detected multiple crashes in _wazuh-db_. We do not have a core dump yet, but we have some suspicions.

### Clues

- It's produced by an agent on Windows.
- It's related to FIM, probably a Registry item.
- ~The agent version is 4.4.x.~

**Comment by TomasTurina** [comment]
## Update

The cloud team shared several core dumps with us, none of which could be inspected to determine where the failure was. For this reason we proceeded to analyze the logs, which was the only thing we had at our disposal. At the same time we were trying to reproduce the bug locally without success.

From the Wazuh DB debug, we were able to identify that the only two modules that were making queries to Wazuh DB were FIM and Syscollector, so we intuited that one of them could be the cause of the failure. After identifying this, we ask the support team to disable these two modules in order to confirm our suspicions and have a point to continue the investigation.

No particular pattern was found in the logs that causes Wazuh DB to crash. Although `fim_registry` and `dbsync` type queries can be observed in the previous moments, this is not something that we can confirm yet as the cause of the problem.

Another thing that we proposed yesterday was to test locally generated packages in debug to see if it was possible to read these core dumps, but we were not successful with this either.

We have seen the failure only with Windows agents. We originally thought that the bug occurred when agents with version `4.4.x` tried to connect to a `4.3.x` manager. However, we rule out that it is a problem of this type because the cloud team confirmed to us that there are environments with only `4.3.x` agents with the problem (they have Windows agents).

We agreed late yesterday with the support team to disable FIM and Syscollector in the analyzed cloud environment and this made the problem go away, further confirming our suspicions about FIM and Windows agents. On the other hand, cloud proceeded to activate only Syscollector and the problem did not occur either, so this last module can almost certainly be ruled out as the source of the problem.

Additionally we were doing tests in local environments. One of them consisted of executing all the queries that were in the log using the `wdb-query.py` tool. The result of this was that none of the queries generated any crashes, so we rule out that any particular query is the cause of the crash but rather that it seems to be a particular condition (perhaps a race condition) leading to the problem.

We also saw that the error that Wazuh DB was throwing when crashing came from a call to the` abort()` function, which can be linked to several reasons of which we suspect two:
- A double free of a pointer.
- An assert that is not true, which can happen because cloud packages are compiled with the debug flag.

We look forward to future core dumps that can be read and analyzed to continue this research.

**Comment by Dwordcito** [comment]
Last logs before the sigabort in the last wazuh-db stop.

```
2023/04/18 14:28:37 wazuh-db[331] wdb_parser.c:1345 at wdb_parse_syscollector(): DEBUG: DB(298) syscollector_processes Syscollector query. 
2023/04/18 14:28:37 wazuh-db[331] wdb_parser.c:245 at wdb_parse(): DEBUG: Agent 222 query: fim_registry save2 {"xxx"}}
2023/04/18 14:28:37 wazuh-db[331] wdb_parser.c:245 at wdb_parse(): DEBUG: Agent 222 query: fim_registry save2 {"xxx"}}
2023/04/18 14:28:37 wazuh-db[331] wdb_parser.c:245 at wdb_parse(): DEBUG: Agent 222 query: fim_registry save2 {"xxx"}}
2023/04/18 14:28:37 wazuh-db[331] wdb_parser.c:245 at wdb_parse(): DEBUG: Agent 145 query: fim_registry save2 {"xxx"}}
```



GDB
```
isplaying notes found at file offset 0x000012d8 with length 0x0000a3a0:
  Owner                Data size 	Description
  CORE                 0x00000150	NT_PRSTATUS (prstatus structure)	
  CORE                 0x00000088	NT_PRPSINFO (prpsinfo structure)	
  CORE                 0x00000080	NT_SIGINFO (siginfo_t data)	
  CORE                 0x00000140	NT_AUXV (auxiliary vector)	
  CORE                 0x0000062a	NT_FILE (mapped files)	    Page size: 4096
                 Start                 End         Page Offset
    0x0000000000400000  0x0000000000496000  0x0000000000000000
        /bin/wazuh-db
    0x0000000000696000  0x0000000000698000  0x0000000000000096
        /bin/wazuh-db
    0x00007fb1e7d35000  0x00007fb1e7d41000  0x0000000000000000
        /usr/lib64/libnss_files-2.17.so
    0x00007fb1e7d41000  0x00007fb1e7f40000  0x000000000000000c
        /usr/lib64/libnss_files-2.17.so
    0x00007fb1e7f40000  0x00007fb1e7f41000  0x000000000000000b
        /usr/lib64/libnss_files-2.17.so
    0x00007fb1e7f41000  0x00007fb1e7f42000  0x000000000000000c
        /usr/lib64/libnss_files-2.17.so
    0x00007fb1e7f48000  0x00007fb1e810c000  0x0000000000000000
        /usr/lib64/libc-2.17.so
    0x00007fb1e810c000  0x00007fb1e830b000  0x00000000000001c4
        /usr/lib64/libc-2.17.so
    0x00007fb1e830b000  0x00007fb1e830f000  0x00000000000001c3
        /usr/lib64/libc-2.17.so
    0x00007fb1e830f000  0x00007fb1e8311000  0x00000000000001c7
        /usr/lib64/libc-2.17.so
    0x00007fb1e8316000  0x00007fb1e832d000  0x0000000000000000
        /usr/lib64/libpthread-2.17.so
    0x00007fb1e832d000  0x00007fb1e852c000  0x0000000000000017
        /usr/lib64/libpthread-2.17.so
    0x00007fb1e852c000  0x00007fb1e852d000  0x0000000000000016
        /usr/lib64/libpthread-2.17.so
    0x00007fb1e852d000  0x00007fb1e852e000  0x0000000000000017
        /usr/lib64/libpthread-2.17.so
    0x00007fb1e8532000  0x00007fb1e8633000  0x0000000000000000
        /usr/lib64/libm-2.17.so
    0x00007fb1e8633000  0x00007fb1e8832000  0x0000000000000101
        /usr/lib64/libm-2.17.so
    0x00007fb1e8832000  0x00007fb1e8833000  0x0000000000000100
        /usr/lib64/libm-2.17.so
    0x00007fb1e8833000  0x00007fb1e8834000  0x0000000000000101
        /usr/lib64/libm-2.17.so
    0x00007fb1e8834000  0x00007fb1e90d3000  0x0000000000000000
        /lib/libwazuhext.so
    0x00007fb1e90d3000  0x00007fb1e92d2000  0x000000000000089f
        /lib/libwazuhext.so
    0x00007fb1e92d2000  0x00007fb1e932e000  0x000000000000089e
        /lib/libwazuhext.so
    0x00007fb1e9355000  0x00007fb1e9357000  0x0000000000000000
        /usr/lib64/libdl-2.17.so
    0x00007fb1e9357000  0x00007fb1e9557000  0x0000000000000002
        /usr/lib64/libdl-2.17.so
    0x00007fb1e9557000  0x00007fb1e9558000  0x0000000000000002
        /usr/lib64/libdl-2.17.so
    0x00007fb1e9558000  0x00007fb1e9559000  0x0000000000000003
        /usr/lib64/libdl-2.17.so
    0x00007fb1e9559000  0x00007fb1e9560000  0x0000000000000000
        /usr/lib64/librt-2.17.so
    0x00007fb1e9560000  0x00007fb1e975f000  0x0000000000000007
        /usr/lib64/librt-2.17.so
    0x00007fb1e975f000  0x00007fb1e9760000  0x0000000000000006
        /usr/lib64/librt-2.17.so
    0x00007fb1e9760000  0x00007fb1e9761000  0x0000000000000007
        /usr/lib64/librt-2.17.so
    0x00007fb1e9761000  0x00007fb1e9783000  0x0000000000000000
        /usr/lib64/ld-2.17.so
    0x00007fb1e9982000  0x00007fb1e9983000  0x0000000000000021
        /usr/lib64/ld-2.17.so
    0x00007fb1e9983000  0x00007fb1e9984000  0x0000000000000022
        /usr/lib64/ld-2.17.so
```

GDB
```
Core was generated by `/var/ossec/bin/wazuh-db'.
Program terminated with signal SIGABRT, Aborted.
#0  0x00007fb1e7f7e387 in ?? ()
```
Text segment from shared library.
    0x00007fb1e7f48000  0x00007fb1e810c000  0x0000000000000000
        /usr/lib64/libc-2.17.so

0x00007fb1e7f7e387 - 0x00007fb1e7f48000 = 36387

```
[root@d90980e14c0b /]# objdump -d -M x86-64 libc-2.17.so | grep 36387 -A 18 -B 18
0000000000036350 <raise>:
   36350:	64 8b 0c 25 d4 02 00 	mov    %fs:0x2d4,%ecx
   36357:	00 
   36358:	64 8b 34 25 d0 02 00 	mov    %fs:0x2d0,%esi
   3635f:	00 
   36360:	85 f6                	test   %esi,%esi
   36362:	75 34                	jne    36398 <raise+0x48>
   36364:	b8 ba 00 00 00       	mov    $0xba,%eax
   36369:	0f 05                	syscall 
   3636b:	89 c1                	mov    %eax,%ecx
   3636d:	64 89 04 25 d0 02 00 	mov    %eax,%fs:0x2d0
   36374:	00 
   36375:	89 c6                	mov    %eax,%esi
   36377:	48 63 d7             	movslq %edi,%rdx
   3637a:	48 63 f6             	movslq %esi,%rsi
   3637d:	48 63 f9             	movslq %ecx,%rdi
   36380:	b8 ea 00 00 00       	mov    $0xea,%eax
   36385:	0f 05                	syscall 
   36387:	48 3d 00 f0 ff ff    	cmp    $0xfffffffffffff000,%rax
   3638d:	77 1e                	ja     363ad <raise+0x5d>
   3638f:	f3 c3                	repz retq 
   36391:	0f 1f 80 00 00 00 00 	nopl   0x0(%rax)
   36398:	85 c9                	test   %ecx,%ecx
   3639a:	7f db                	jg     36377 <raise+0x27>
   3639c:	89 c8                	mov    %ecx,%eax
   3639e:	f7 d8                	neg    %eax
   363a0:	81 e1 ff ff ff 7f    	and    $0x7fffffff,%ecx
   363a6:	0f 44 c6             	cmove  %esi,%eax
   363a9:	89 c1                	mov    %eax,%ecx
   363ab:	eb ca                	jmp    36377 <raise+0x27>
   363ad:	48 8b 15 9c 0a 39 00 	mov    0x390a9c(%rip),%rdx        # 3c6e50 <.got+0x100>
   363b4:	f7 d8                	neg    %eax
   363b6:	64 89 02             	mov    %eax,%fs:(%rdx)
   363b9:	48 83 c8 ff          	or     $0xffffffffffffffff,%rax
   363bd:	c3                   	retq   
   363be:	66 90                	xchg   %ax,%ax
```
this concludes that the abort is correct.

```
[Current thread is 1 (LWP 338)]
(gdb) info registers
rax            0x0                 0
rbx            0x7fb1e4b0dce0      140402022735072
rcx            0x7fb1e7f7e387      140402077721479
rdx            0x6                 6
rsi            0x152               338
rdi            0x14a               330
rbp            0x7fb1e4b0e3c0      0x7fb1e4b0e3c0
rsp            0x7fb1e4b0d988      0x7fb1e4b0d988
r8             0x0                 0
r9             0x7fb1dc0008c0      140401876928704
r10            0x8                 8
r11            0x202               514
r12            0x0                 0
r13            0x7                 7
r14            0x62                98
r15            0x7                 7
rip            0x7fb1e7f7e387      0x7fb1e7f7e387
eflags         0x202               [ IF ]
cs             0x33                51
ss             0x2b                43
ds             0x0                 0
es             0x0                 0
fs             0x0                 0
gs             0x0                 0
k0             0x0                 0
k1             0x0                 0
k2             0x0                 0
k3             0x0                 0
k4             0x0                 0
k5             0x0                 0
k6             0x0                 0
k7             0x0                 0

next frame->

(gdb) x/gx $rbp+8
0x7fb1e4b0e3c8:	0x00007fb1e8ba4b1c
```
0x00007fb1e8834000  0x00007fb1e90d3000  0x0000000000000000
        /lib/libwazuhext.so
        

0x00007fb1e8ba4b1c - 0x00007fb1e8834000 = 370B1C

0000000000370b00 <sqlite3MemMalloc>:
  370b00:	48 89 6c 24 f8       	mov    %rbp,-0x8(%rsp)
  370b05:	48 63 ef             	movslq %edi,%rbp
  370b08:	48 89 5c 24 f0       	mov    %rbx,-0x10(%rsp)
  370b0d:	8d 7d 08             	lea    0x8(%rbp),%edi
  370b10:	48 83 ec 18          	sub    $0x18,%rsp
  370b14:	48 63 ff             	movslq %edi,%rdi
  370b17:	e8 8c 90 da ff       	call   119ba8 <malloc@plt>
  370b1c:	48 85 c0             	test   %rax,%rax
  370b1f:	48 89 c3             	mov    %rax,%rbx
  370b22:	74 1c                	je     370b40 <sqlite3MemMalloc+0x40>
  370b24:	48 89 28             	mov    %rbp,(%rax)
  370b27:	48 83 c3 08          	add    $0x8,%rbx
  370b2b:	48 89 d8             	mov    %rbx,%rax
  370b2e:	48 8b 6c 24 10       	mov    0x10(%rsp),%rbp
  370b33:	48 8b 5c 24 08       	mov    0x8(%rsp),%rbx
  370b38:	48 83 c4 18          	add    $0x18,%rsp
  370b3c:	c3                   	ret    
  370b3d:	0f 1f 00             	nopl   (%rax)
  370b40:	48 8d 35 59 00 3b 00 	lea    0x3b0059(%rip),%rsi        # 720ba0 <last_valueName+0x7b0>
  370b47:	89 ea                	mov    %ebp,%edx
  370b49:	bf 07 00 00 00       	mov    $0x7,%edi
  370b4e:	31 c0                	xor    %eax,%eax
  370b50:	e8 43 b4 db ff       	call   12bf98 <sqlite3_log@plt>
  370b55:	eb d4                	jmp    370b2b <sqlite3MemMalloc+0x2b>
  370b57:	66 0f 1f 84 00 00 00 	nopw   0x0(%rax,%rax,1)
  370b5e:	00 00
  370b3d:	0f 1f 00             	nopl   (%rax)
  370b40:	48 8d 35 59 00 3b 00 	lea    0x3b0059(%rip),%rsi        # 720ba0 <sqlite3_version@@Base

**Comment by Dwordcito** [comment]
invalid Basepointer

```
(gdb) x/gx $rbp
0x7fb1e4b0e3c0:	0x0000000000000b98
```

**Comment by Dwordcito** [comment]
backtraces from GDB execution

(gdb) bt
#0  0x00007ffff65f8387 in raise () from /lib64/libc.so.6
#1  0x00007ffff65f9a78 in abort () from /lib64/libc.so.6
#2  0x00007ffff663af67 in __libc_message () from /lib64/libc.so.6
#3  0x00007ffff6643329 in _int_free () from /lib64/libc.so.6
#4  0x000000000041395d in wdb_modify_dbsync (wdb=0x7fffe0186dc0, kv_value=0x46b868 <TABLE_MAP+3752>, data=0x7ffff599cef4 "2023/04/18 16:26:06") at wazuh_db/wdb_global.c:1487
#5  0x000000000042a8da in process_dbsync_data (wdb=0x7fffe0186dc0, kv_value=0x46b868 <TABLE_MAP+3752>, operation=0x7ffff599ceeb "MODIFIED", data=0x7ffff599cef4 "2023/04/18 16:26:06")
    at wazuh_db/wdb_parser.c:5574
#6  0x000000000042aaf9 in wdb_parse_dbsync (wdb=0x7fffe0186dc0, input=0x7ffff599cee1 "processes", output=0x7ffff598cec0 "") at wazuh_db/wdb_parser.c:5619
#7  0x000000000041761d in wdb_parse (input=0x7ffff599ced0 "agent", output=0x7ffff598cec0 "", peer=116) at wazuh_db/wdb_parser.c:439
#8  0x0000000000405f66 in run_worker (args=0x0) at wazuh_db/main.c:366
#9  0x00007ffff6997ea5 in start_thread () from /lib64/libpthread.so.0
#10 0x00007ffff66c0b0d in clone () from /lib64/libc.so.6
[14:50](https://wazuh-team.slack.com/archives/D01TA9KED37/p1681840202347539)
(gdb) bt
#0  0x00007ffff65f8387 in raise () from /lib64/libc.so.6
#1  0x00007ffff65f9a78 in abort () from /lib64/libc.so.6
#2  0x00007ffff663af67 in __libc_message () from /lib64/libc.so.6
#3  0x00007ffff6643329 in _int_free () from /lib64/libc.so.6
#4  0x000000000041395d in wdb_modify_dbsync (wdb=0x7fffcf034570, kv_value=0x46b868 <TABLE_MAP+3752>, data=0x7fffed7eaef4 "2023/04/18 17:27:22") at wazuh_db/wdb_global.c:1487
#5  0x000000000042a8da in process_dbsync_data (wdb=0x7fffcf034570, kv_value=0x46b868 <TABLE_MAP+3752>, operation=0x7fffed7eaeeb "MODIFIED", data=0x7fffed7eaef4 "2023/04/18 17:27:22")
    at wazuh_db/wdb_parser.c:5574
#6  0x000000000042aaf9 in wdb_parse_dbsync (wdb=0x7fffcf034570, input=0x7fffed7eaee1 "processes", output=0x7fffed7daec0 "") at wazuh_db/wdb_parser.c:5619
#7  0x000000000041761d in wdb_parse (input=0x7fffed7eaed0 "agent", output=0x7fffed7daec0 "", peer=57) at wazuh_db/wdb_parser.c:439
#8  0x0000000000405f66 in run_worker (args=0x0) at wazuh_db/main.c:366
#9  0x00007ffff6997ea5 in start_thread () from /lib64/libpthread.so.0
#10 0x00007ffff66c0b0d in clone () from /lib64/libc.so.6
[14:50](https://wazuh-team.slack.com/archives/D01TA9KED37/p1681840243131259)
(gdb) bt
#0  0x00007ffff65f8387 in raise () from /lib64/libc.so.6
#1  0x00007ffff65f9a78 in abort () from /lib64/libc.so.6
#2  0x00007ffff663af67 in __libc_message () from /lib64/libc.so.6
#3  0x00007ffff6644b36 in _int_malloc () from /lib64/libc.so.6
#4  0x00007ffff664778c in malloc () from /lib64/libc.so.6
#5  0x00007ffff721eb1c in sqlite3MemMalloc () from /var/ossec/bin/../lib/libwazuhext.so
#6  0x00007ffff7209085 in sqlite3Malloc () from /var/ossec/bin/../lib/libwazuhext.so
#7  0x00007ffff7209146 in sqlite3MallocZero () from /var/ossec/bin/../lib/libwazuhext.so
#8  0x00007ffff720917d in sqlite3BitvecCreate () from /var/ossec/bin/../lib/libwazuhext.so
#9  0x00007ffff7238bde in pager_write () from /var/ossec/bin/../lib/libwazuhext.so
#10 0x00007ffff723d2d1 in btreeOverwriteContent () from /var/ossec/bin/../lib/libwazuhext.so
#11 0x00007ffff723d422 in btreeOverwriteCell () from /var/ossec/bin/../lib/libwazuhext.so
---Type <return> to continue, or q <return> to quit---
#12 0x00007ffff72422c5 in sqlite3BtreeInsert () from /var/ossec/bin/../lib/libwazuhext.so
#13 0x00007ffff725fae9 in sqlite3VdbeExec () from /var/ossec/bin/../lib/libwazuhext.so
#14 0x00007ffff7266295 in sqlite3_step () from /var/ossec/bin/../lib/libwazuhext.so
#15 0x0000000000408811 in wdb_step (stmt=0x7fffd1c74fb8) at wazuh_db/wdb.c:615
#16 0x0000000000413921 in wdb_modify_dbsync (wdb=0x7fffd035e230, kv_value=0x46b868 <TABLE_MAP+3752>, data=0x7ffff4f9bef4 "2023/04/18 16:26:04") at wazuh_db/wdb_global.c:1483
#17 0x000000000042a8da in process_dbsync_data (wdb=0x7fffd035e230, kv_value=0x46b868 <TABLE_MAP+3752>, operation=0x7ffff4f9beeb "MODIFIED", data=0x7ffff4f9bef4 "2023/04/18 16:26:04")
    at wazuh_db/wdb_parser.c:5574
#18 0x000000000042aaf9 in wdb_parse_dbsync (wdb=0x7fffd035e230, input=0x7ffff4f9bee1 "processes", output=0x7ffff4f8bec0 "") at wazuh_db/wdb_parser.c:5619
#19 0x000000000041761d in wdb_parse (input=0x7ffff4f9bed0 "agent", output=0x7ffff4f8bec0 "", peer=37) at wazuh_db/wdb_parser.c:439
#20 0x0000000000405f66 in run_worker (args=0x0) at wazuh_db/main.c:366
#21 0x00007ffff6997ea5 in start_thread () from /lib64/libpthread.so.0
#22 0x00007ffff66c0b0d in clone () from /lib64/libc.so.6


RCA:

Agents 4.4 are using dead code from wazuh 4.3.10 manager. It is incorrect to use a later version agent with a manager. It is an unproven case and something that can give unexpected behaviors.


Fix: Remove decoder of dbsync from 4.3.x branch.
